### PR TITLE
Remove the incorrect module_id setting

### PR DESCRIPTION
**Pull request by yuanwu2017**
# What does this PR do?
Fix the incorrect model_id issue.
The environment variables have been correctly set in the` initialize_distributed_hpu()` function. No need to 
configure the HLS_MODULE_ID again. 

The following line is wrong. 
`os.environ["HLS_MODULE_ID"] = str(local_rank)`
The right configuration should be like the following. 
```
visible_modules = os.environ[HABANA_VISIBLE_MODULES_VAR].split(",")
os.environ[HLS_MODULE_ID_VAR] = visible_modules[local_rank]
```

Pytest:

image

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/accelerate/state.py [hunk 1]
```diff
@@ -48,7 +48,10 @@ def __init__(self, cpu: bool = False, **kwargs):
             from habana_frameworks.torch.distributed.hccl import initialize_distributed_hpu
 
             if int(os.environ.get("LOCAL_RANK", -1)) != -1 and not cpu:
-                world_size, rank, local_rank = initialize_distributed_hpu()
+                world_size = os.environ.get("WORLD_SIZE", None)
+                local_rank = os.environ.get("LOCAL_RANK", None)
+                rank = os.environ.get("RANK", None)
```
Comment (libinta) [inline optimum/habana/accelerate/state.py:53]: @yuanwu2017 can you point out where these are set in the mpi and deepspeed case? I only remember deepspeed has these set, can you conform?
what test have you done for this?

Comment (yuanwu2017) [reply optimum/habana/accelerate/state.py:53]: 1. https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling#using-deepspeed
image

2. https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling#single-card-training-gpt2
image
3.https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling#multi-card-training-with-deepspeed-gpt-j
image

4.https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling#multi-card-training-gpt2
image

**Comment by imangohari1** [comment]
@yuanwu2017 
Could you please sycn this branch on top of head/main  and rebase it. 
are there any impact on any model/examples with `--use_mpi` with these changes?

File: optimum/habana/accelerate/state.py [hunk 2]
```diff
@@ -60,10 +60,6 @@ def __init__(self, cpu: bool = False, **kwargs):
                     self.distributed_type = GaudiDistributedType.DEEPSPEED
                     import deepspeed
 
-                    if world_size > 1:
```
Comment (imangohari1) [inline optimum/habana/accelerate/state.py:63]: These changes were added here: 
https://github.com/huggingface/optimum-habana/pull/1089/files

let's repeat your experiments in virtual env and see if indeed these are not needed. 
if the changes are not the best coded apprach, pls change them and retest.

**Review by imangohari1 — CHANGES_REQUESTED** [review]
Hi @yuanwu2017 
We need to test these in virtual env. 
Pls. see me comment below. 
Thanks.

File: optimum/habana/accelerate/state.py (hunk 2 above)
Comment (yuanwu2017) [reply optimum/habana/accelerate/state.py:63]: Ok. Thanks.

**Comment by yuanwu2017** [comment]
> @yuanwu2017 Could you please sycn this branch on top of head/main and rebase it. are there any impact on any model/examples with `--use_mpi` with these changes?

Testcase 4 is for distributed training with use_mpi.
```
python ../gaudi_spawn.py \
    --world_size 8 --use_mpi run_clm.py \
    --model_name_or_path gpt2 \
    --dataset_name wikitext \
    --dataset_config_name wikitext-2-raw-v1 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --do_train \
    --do_eval \
    --output_dir /tmp/test-clm \
    --gaudi_config_name Habana/gpt2 \
    --use_habana \
    --use_lazy_mode \
    --use_hpu_graphs_for_inference \
    --gradient_checkpointing \
    --use_cache False \
    --throughput_warmup_steps 3
```

**Comment by imangohari1** [comment]
@regisss 
A review on this would be appreciated. Thanks.

**Comment by imangohari1** [comment]
@yuanwu2017 
Let us know when you tested this PR in Virtual environment. 
Meanwhile I suggest putting in WIP or Draft. Thanks

**Comment by yuanwu2017** [comment]
> @yuanwu2017 Let us know when you tested this PR in Virtual environment. Meanwhile I suggest putting in WIP or Draft. Thanks

The above tests were all tested in the docker environment. Can you tell me which cases still need to be tested?

**Comment by imangohari1** [comment]
> > @yuanwu2017 Let us know when you tested this PR in Virtual environment. Meanwhile I suggest putting in WIP or Draft. Thanks
> 
> The above tests were all tested in the docker environment. Can you tell me which cases still need to be tested?

I would suggest:
- running few tests with KVM setup in our internal clusters
- with limiting the #HPU visible in the docker container. 
  - Please redo the tests conducted [here](https://github.com/huggingface/optimum-habana/pull/1089#issue-2367852400) 

Thanks

**Comment by yuanwu2017** [comment]
`sudo hl-smi -Q index,module_id -f csv`
image
**Docker command 1:**

> `docker run -it -p 8083:80 -v ~/workspace/data:/data -v ~/workspace/tmp:/tmp -v ~/workspace:/workspace --runtime=habana --name tgi-yuanwu -e PT_HPU_ENABLE_LAZY_COLLECTIVES=true -e HABANA_VISIBLE_DEVICES=4,6 -e HABANA_VISIBLE_MODULES=4,0 -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_nice --ipc=host tgi:2.0.4`
> 
> **Command:**
> ` python ../gaudi_spawn.py --world_size 2 --use_deepspeed run_clm.py --model_name_or_path gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --per_device_train_batch_size 4 --per_device_eval_batch_size 4 --do_train --do_eval --output_dir /tmp/test-clm --gaudi_config_name Habana/gpt2 --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --throughput_warmup_steps 3 --deepspeed deepspeed_zero_2.json`
> **Result:**
> image
> The processes ran on device 4 and 6.

**Docker command 2:**
`docker run -it -p 8083:80 -v ~/workspace/data:/data -v ~/workspace/tmp:/tmp -v ~/workspace:/workspace --runtime=habana --name tgi-yuanwu -e PT_HPU_ENABLE_LAZY_COLLECTIVES=true -e HABANA_VISIBLE_DEVICES=6,7  -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_nice --ipc=host tgi:2.0.4`

> **result:**
image
The processes ran on the correct device 6 and 7.

**Comment by yuanwu2017** [comment]
* running few tests with KVM setup in our internal clusters
I ran these tests on internal habana cluster with docker container.  Do you mean I should run tests without docker container?

**Comment by yuanwu2017** [comment]
@imangohari1 
Run the test on bare metal without Docker container.
image
**### Run test on device 0 and 1**
**Command:** 
```
PT_HPU_ENABLE_LAZY_COLLECTIVES=true \
HABANA_VISIBLE_DEVICES=0,1 \
HABANA_VISIBLE_MODULES=2,6 \
exportOMPI_MCA_btl_vader_single_copy_mechanism=none \
python ../gaudi_spawn.py --world_size 2 --use_deepspeed run_clm.py --model_name_or_path gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --per_device_train_batch_size 4 --per_device_eval_batch_size 4 --do_train --do_eval --output_dir ./test-clm --gaudi_config_name Habana/gpt2 --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --throughput_warmup_steps 3 --deepspeed deepspeed_zero_2.json
```
Result:
image

Use the correct devices.
image

**### Run test on device 1 and 3**
**Command:** 
```
PT_HPU_ENABLE_LAZY_COLLECTIVES=true \
HABANA_VISIBLE_DEVICES=1,3 \
HABANA_VISIBLE_MODULES=6,3 \
exportOMPI_MCA_btl_vader_single_copy_mechanism=none \
python ../gaudi_spawn.py --world_size 2 --use_deepspeed run_clm.py --model_name_or_path gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --per_device_train_batch_size 4 --per_device_eval_batch_size 4 --do_train --do_eval --output_dir ./test-clm --gaudi_config_name Habana/gpt2 --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --throughput_warmup_steps 3 --deepspeed deepspeed_zero_2.json
```
image

Use the correct devices.
image

**Summary:**
When running distributed training on bare metal without docker container. Both HABANA_VISIBLE_DEVICES and HABANA_VISIBLE_MODULES must be set. Because nobody help to set the HABANA_VISIBLE_MODULES.
When running distributed training in docker container, only HABANA_VISIBLE_DEVICES  must be set. Because docker will help to set the HABANA_VISIBLE_MODULES when setting up the container. But what needs to be careful is that their relationship is not exactly one-to-one like command“sudo hl-smi -Q index,module_id -f csv”. His corresponding relationship may be wrong, but using those module ids is accurate.
image

**Comment by yuanwu2017** [comment]
Close the pull request. The issue has been fixed.

**Pull request closed without merging**